crates/ra_lsp_server/src/main_loop.rs

```diff
@@ -578,8 +578,13 @@ fn on_notification(
         Err(not) => not,
     };
     let not = match notification_cast::<req::DidSaveTextDocument>(not) {
-        Ok(_params) => {
+        Ok(params) => {
             state.check_watcher.update();
+            let uri = params.text_document.uri;
+            let path = uri.to_file_path().map_err(|()| format!("invalid uri: {}", uri))?;
+            if let Some(file_id) = state.vfs.write().remove_file_overlay(&path) {
```

Comment:
Hm, I am pretty sure this should only happen in `DidCloseTextDocument`. The editor still can send us incremental changes after DidSave event

Reply:
You probably know that better than me :)

As far as I understand the code we have a few ways of re-scheduling diagnostics here:
1. Send some sort of VFS event on the file, which schedules an update
2. Change the invalidation login in loop_turn to allow us to just add a subscription and redo the diagnostics.
3. Pass the thread pool down into the event function and schedule an update of the specific file there, like we do when receiving a cargo watch.

Do you have a preference here?

Reply:
It is correct that this will probably lead to diagnostics rescheduling, but it will also remove the file from in-memory vfs, which we don't want to do :)

But why do we want to re-trigger diagnostics on save *here* at all? It seems like we should re-trigger then once we get a message from check_watcher, and we'll get it because we call `state.check_watcher.update();`.

That is, the previous logic here seems entirely corrent?

Reply:
We don't know what file to clear/update at that point. And we lose that information once we cancel the update that was supposed to clear the diagnostics in the first place.

Reply:
But we don't clear diagnostics when files are modified, we clear them when we receive Clear tasks.

So my understanding is that the effect of this change would be:

* on save we flag file as modified.
* we schedule diagnostics
* we schedule cargo check update
* diagnostics are run, using the old state of check diagnostics
* clear event comes from cargo check, we clear diagnostics and schedule handle_diagnostics again

Reply:
That seems like a correct understanding yes. The key issue I'm trying to solve, albeit in quite aggressive way, is that: If we get canceled during updating diagnostics, as response to having received clear event, we lose the information about which files to clear. This results in stale diagnostics if the started cargo run returns no new diagnostics (e.g on success).
There's two main ways I can think of to fix this:
* Update diagnostics on a user event the user would expect to clear stale diagnostics (e.g. on save)
* Make the diagnostic update not cancelable when we are clearing old cargo diagnostics, but I'm not sure if we support this at all.

Reply:
But we still only clear diagnostics in the backgound thread, so it still can get cancelled. 

I am not sure there's an easier way to solve this than just making sure that `cargo check` and "native" daignostics use different code paths, and are merged in the main loop explicitly.

Reply:
Fair enough, I will give the suggested design a look, and try it out once I find some time during the day :+1: